quanto/tensor/core.py

```diff
@@ -248,7 +337,14 @@ def forward(ctx, base, itype, axis=None):
         scale = (rmax - rmin) / (qmax - qmin)
         zeropoint = torch.round(-rmin / scale).to(torch.int8)
         data = torch.clamp(torch.round((base - rmin) / scale), min=0, max=2**bits - 1).to(torch.int8)
+
+        # pack the weights only if the weights are unpacked
+        can_pack_weights = data.shape[0] // 32 * itype.bits != 0
```

Comment:
True, we can definitely flatten the tensor and retrieve the original shape during unpacking, let me try to do that based on your commit you shared and get back

Reply:
Just realized I forgot to mask when unpacking.
pack -> shift, then or
unpack -> mask, then unshift

Reply:
In the light of some painful attempts at implementing something, I am reconsidering some of my comments:

1. Pack to 8-bit, not 32-bit
In order to use pytorch operations (and thus avoid moving data from the device ot the cpu), we cannot pack to 32 bits since `torch.uint32` is only in 2.2. beta, without any supported op. We therefore must pack to `torch.uint8`, the only supported unsigned type.
Note that this helps also for point 2, since if we pack along a single dim, we only need to make sure the dim is a multiple of 2/4.

2. Preserve the original shape and pack along the first dim
I finally understood why the other implementations pack along the first dimension: it is because it makes pack/unpack so much more efficient (and by that I mean fast).
Since tensors are row-major, and the first dim corresponds to the output feature, each output feature weights correspond to a contiguous buffer. This means that slicing along the first dim is very efficient: it is just a jump to the buffer corresponding to the next group of output features.
This means that when packing we can split the original tensor in 2/4, shift the splits element-wise and add them.
Conversely, when unpacking, we can mask + unshift the packed tensor 2/4 times, then simply cat the results.
This also means that when fusing the unpack with a mm, we can fully parallelize (each output feature being independent each split can be processed independently).
You can see an example in HQQ.

Reply:
Sounds great! Sorry for the delay replying on this, I will work on the points you raised here